apisix/cli/ops.lua

```diff
@@ -807,17 +807,19 @@ local function start(env, ...)
         local local_conf_path = profile:yaml_path("config")
         local local_conf_path_bak = local_conf_path .. ".bak"
 
-        local ok, err = os_rename(local_conf_path, local_conf_path_bak)
-        if not ok then
-            util.die("failed to backup config, error: ", err)
-        end
-        local ok, err1 = lfs.link(customized_yaml, local_conf_path)
-        if not ok then
-            ok, err = os_rename(local_conf_path_bak,  local_conf_path)
+        if not pl_path.exists(local_conf_path_bak) then
+            local ok, err = os_rename(local_conf_path, local_conf_path_bak)
             if not ok then
-                util.die("failed to recover original config file, error: ", err)
+                util.die("failed to backup config, error: ", err)
+            end
+            local ok, err1 = lfs.link(customized_yaml, local_conf_path)
+            if not ok then
+                ok, err = os_rename(local_conf_path_bak,  local_conf_path)
+                if not ok then
+                    util.die("failed to recover original config file, error: ", err)
+                end
+                util.die("failed to link customized config, error: ", err1)
             end
-            util.die("failed to link customized config, error: ", err1)
         end
```

Comment:
Is this an appropriate change? the customized_yaml won't work if there is a `local_conf_path_bak` file

Reply:
Indeed I have another option, why do we need a backup file?

Reply:
Once start APISIX use customized config, the original `config.yaml ` will be `rename` to `config.yaml.bak`  and set a hard link from customized config to `config.yaml`, if occur some errors cause start APISIX fail then it will not restore these config files, so we don't need do above opeartion again, just use `config.yaml` as apisix config directly is enough.
Unless the user create `local_conf_path_bak` manually before start apisix. 
If we can detected if have errors when start APISIX, it maybe the best solution, but, I have no idea about this.

Reply:
I think we can fix this bug by https://github.com/apache/apisix/pull/9701#discussion_r1259432308
Do the `cleanup` when abnormal

Reply:
I fetch codes of this PR, and test it, it can't resolve this https://github.com/apache/apisix/issues/9171. You can try it.

Reply:
I think @tokers is right, we should find a way to access customized config file directly, but not rename the customized config file to origin `config.yaml`, it brings too many problems.

I have a simple idea, we can save the custom config file name to a sparate file(like `.config_name`) when we start apisix, so we can access it directly when we need it.

Reply:
Then you will have two config file: config.yaml and the custom_config.yaml, how do you decide which one to choose each time you start APISIX?

Reply:
The pr is not right, what i want to say is that *when starting APISIX fails, recover the config.yaml file*.  
So we can do like this:
```
    if env.deployment_role ~= "data_plane" then
        local err = init_etcd(env, args)
        if err then
            cleanup()
            util.die(err)
        end
    end
```
Now we die in the `init_etcd` with no returns, so we should modify the `init_etcd` function to make sure the it returns the error msg instead of just crashing!

Reply:
> Then you will have two config file: config.yaml and the custom_config.yaml, how do you decide which one to choose each time you start APISIX?

which config will be choose depend on `--config` flag.  After APISIX started, we will read `.config_name` file content to decide which config will be choose.

Reply:
> The pr is not right, what i want to say is that _when starting APISIX fails, recover the config.yaml file_. We can do like this:
> 
> ```
>     if env.deployment_role ~= "data_plane" then
>         local err = init_etcd(env, args)
>         if err then
>             cleanup()
>             util.die(err)
>         end
>     end
> ```
> 
> Now we die in the `init_etcd` with no returns, so we should modify the `init_etcd` function to make sure the it returns the error msg instead of just crashing! The key point is `cleanup()`

I'm not sure if only `init_etcd` will cause crash, if so, I think is ok.

Reply:
> which config will be choose depend on --config flag. After APISIX started, we will read .config_name file content to decide which config will be choose.
Looks good, but you still have to copy the custom config file to the conf directory, right?

Reply:
Emm, In current logic, you are right. Give me some time to test it, I think it can be solved.